presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -438,13 +440,34 @@ protected TupleDescriptor visitInsert(Insert insert, AnalysisContext context)
         accessControl.checkCanInsertIntoTable(session.getIdentity(), targetTable);
         analysis.setInsertTarget(targetTableHandle.get());
 
-        List<ColumnMetadata> columns = metadata.getTableMetadata(session, targetTableHandle.get()).getColumns();
-        Iterable<Type> tableTypes = columns.stream()
-                .filter(column -> !column.isHidden())
-                .map(ColumnMetadata::getType)
-                .collect(toImmutableList());
+        TableMetadata tableMetadata = metadata.getTableMetadata(session, targetTableHandle.get());
+        List<ColumnMetadata> allTableColumns = tableMetadata.getColumns();
 
-        Iterable<Type> queryTypes = transform(descriptor.getVisibleFields(), Field::getType);
+        List<String> insertColumns = insert.getColumns().orElse(tableMetadata.getVisibleColumnNames());
+        analysis.setInsertColumns(insertColumns);
+
+        Set<String> columnNames = new HashSet<>();
+        for (String name : insertColumns) {
```

Comment:
All these validations shouldn't be done if the columns were not listed explicitly. I'd restructure it as:

```java
List<String> insertColumns = tableMetadata.getVisibleColumnNames();
if (insert.getColumns().isPresent()) { 
   ... validate duplicate names
   ... ensure columns exist in target table
}

analysis.setInsertColumns(insertColumns);
... derive types ...
```